blockchain/chainio.go

```diff
@@ -1155,6 +1159,101 @@ func dbPutUtxoView(dbTx database.Tx, view *UtxoViewpoint) error {
 	return nil
 }
 
+// -----------------------------------------------------------------------------
+// The utxo set state contains information regarding the current state of the
+// utxo set.  In particular, it tracks the block height and block hash of the
+// last completed flush.
+//
+// The utxo set state is tracked in the database since at any given time, the
+// utxo cache may not be consistent with the utxo set in the database.  This is
+// due to the fact that the utxo cache only flushes changes to the database
+// periodically.  Therefore, during initialization, the utxo set state is used
+// to identify the last flushed state of the utxo set and it can be caught up
+// to the current best state of the main chain.
+//
+// Note: The utxo set state MUST always be updated in the same database
+// transaction that the utxo set is updated in to guarantee that they stay in
+// sync in the database.
+//
+// The serialized format is:
+//
+//   <block height><block hash>
+//
+//   Field          Type             Size
+//   block height   VLQ              variable
+//   block hash     chainhash.Hash   chainhash.HashSize
+//
+// -----------------------------------------------------------------------------
+
+// utxoSetState represents the current state of the utxo set.  In particular,
+// it tracks the block height and block hash of the last completed flush.
+type utxoSetState struct {
+	lastFlushHeight uint32
+	lastFlushHash   *chainhash.Hash
```

Comment:
This should be a concrete hash instead of a pointer, imo.  The general rule of thumb is to use concrete hashes in structs and take pointers as params.  This help improve cache locality and reduces the number of objects the gC has to track.  It also generally allows prints to display as expected instead of the pointer.

For example, I cause the test to fail on purpose and you can see the output isn't that useful:

```
    TestDbFetchUtxoSetState: src/github.com/decred/dcrd/blockchain/chainio_test.go:1272: "fresh database (no utxo set state saved)": mismatched state:
        want: <nil>
         got: &{lastFlushHeight:0 lastFlushHash:0xc000010b60}
```

Reply:
Makes sense and sounds like a good rule of thumb to me.  I used a concrete hash for `lastFlushHash` in the `UtxoCache` for the same reasons but never updated here.  Updated.